### PR TITLE
Adds Gather Subset Feature to na_ontap_gather_facts

### PR DESCRIPTION
**Pull request by coreywan**
##### SUMMARY

Adds gather_subset argument. (modeled after ios_facts) . It allows a playbook writer to gather just a subset of facts to increase the speed of the task when dealing with a large cluster.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

na_ontap_gather_facts
##### ANSIBLE VERSION

```paste below
ansible 2.7.0
  config file = /Users/wanlessc/innovate/ansible-netapp/ansible.cfg
  configured module search path = [u'/Users/wanlessc/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/wanlessc/miniconda2/envs/ansible2.7/lib/python2.7/site-packages/ansible
  executable location = /Users/wanlessc/miniconda2/envs/ansible2.7/bin/ansible
  python version = 2.7.15 |Anaconda, Inc.| (default, May  1 2018, 18:37:05) [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]

```

##### ADDITIONAL INFORMATION

In my cluster, gathering all the facts took over a minute. When I added this feature and only gathered the aggregate data that I needed, it took it down to 1 second.

```paste below
netapp : Gather All the Facts ---------------------------------------------- 71.49s
/Users/wanlessc/innovate/ansible-netapp/roles/netapp/tasks/test.yaml:12 -----------
netapp : Gather Only Aggregate Facts ---------------------------------------- 0.98s
/Users/wanlessc/innovate/ansible-netapp/roles/netapp/tasks/test.yaml:1 ------------

```

### CODE REVIEW

**Comment by carchi8py** [comment]
Give me a day or 2 to run some test on this and I'll get back to you, but it looks good just quickly looking at the code.

**Review by carchi8py — APPROVED** [review]
shipit

**Comment by jmkleijer** [comment]
Hi Corey,

About 1-2 months ago, I made a similar attempt to limit the gathering of facts as in our (somehwat large) environment, this could take minutes and even then we would not get everything. 
To make it suit our needs a bit more, I’ve decided to allow for the following:
-	Add a parameter called “objects” which tells na_ontap_gather_facts which information we want to retrieve, e.g. 
o	“aggregate_info”
o	“qtree_info”
o	“volume_info”
o	“vserver_info”
You can have multiple “objects” in a single gather_facts run.
-	Add a parameter called “limits”, which, per “object” that you wish to retrieve, limits the objects that you wish to retrieve by putting limits in place by following the basic “query” structure in NMSDK, e.g.: 
o	objects:
  - “volume_info”
limits:
  volume_info:
    query:
      “volume-info”:
        “volume-id-attributes”:
          “owning-vserver-name”: “{{ myvserver }}”
    “desired-attributes”:
      “volume-attributes”:
        “volume-id-attributes”:
          “owning-vserver-name”: “”
          “name”: “”
        “volume-inode-attributes”: “”
As you can see, this follows the same tree like structure that we use when using a plain NMSDK call. Quotes around the keys are of course necessary when the key contains a hyphen. The above example, gives all of the volumes of SVM <myvserver> along with their number of inodes. Since you can have multiple objects in a single gather_facts run, you can also have multiple limits (one per “object”) in a single gather_facts run.
-	As mentioned, in large environments, the gathering of facts would stop halfway through and only return between 600-800 objects of information. I adjusted the code in call_api to do collections of 200 objects each time and keep on running until all items have been fetched. This seems to solve the problem I ran into.

I’m not a programmer by profession so my code will probably not look as clean as it possibly could be. (I’m especially not fond of the way I loop through the query items in call_api as I believe this should be handled by a recursive function call but I’m unsure how to go about this.)

Could this be something?

[nxp_na_ontap_gather_facts.zip](https://github.com/ansible/ansible/files/2639614/nxp_na_ontap_gather_facts.zip)

**Comment by coreywan** [comment]
@jmkleijer ,

Thanks for uploading it. I will try to look at it over the weekend, and see how we can fit it in.

**Comment by jmkleijer** [comment]
Hi Corey,

Thanks for the quick response.
In the last few days I've improved my original code and made it a bit more
flexible and less redundant I think.
(see attached version)
This version does away with all the separate checks / if conditions and
only has a single check in a for-loop where it loops through the keys in
the objects dictionary (if there are any), looks up the appropriate
api-call in a different dictionary and then does the rest. (since the code
was the same for each and every api-call)

Kind regards,

Jeroen Kleijer

**Comment by lonico** [comment]
Hi Corey 

We made a few improvements to gather_facts, and they are now merged into devel.
I was wondering if you're planning to update your PR, so that your changes could be merged as well.

thank you

File: test/units/modules/storage/netapp/test_na_ontap_gather_facts.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def test_ensure_command_called(self, mock_ems_log):
         my_obj = my_module(module)
         my_obj.server = MockONTAPConnection('vserver', 'SVMadmin')
         with pytest.raises(KeyError) as exc:
-            my_obj.get_all()
+            my_obj.get_all('all')
```
Comment (lonico) [inline test/units/modules/storage/netapp/test_na_ontap_gather_facts.py]: it should be 
my_obj.get_all(['all'])

**Review by lonico — CHANGES_REQUESTED** [review]
Thank you for doing this.

File: test/units/modules/storage/netapp/test_na_ontap_gather_facts.py (hunk 1 above)
Comment (coreywan) [reply test/units/modules/storage/netapp/test_na_ontap_gather_facts.py]: @lonico ,

Thanks for spotting this. I fixed it, and I also added some checks for some of the subset stuff.

**Review by lonico — APPROVED** [review]
shipit

and thank you!

**Review by carchi8py — APPROVED** [review]
Ship It

**Review by JohnLieske — APPROVED** [review]
Looks good, tests are good. :+1:

**Review by lonico — COMMENTED** [review]
Corey

Sorry to bug you with this.  

1) There is a conflict because we added new queries for NVMe features.
2) I wonder if it's possible to add a min_version (and maybe a max_version field).   For instance NVMe ZAPIs are only available starting with ONTAP 9.4 (which is ontapi version 140).

We have a fix under review to check for the version, but that would create another conflict.

If this is too much work, let me know, and we could take over and merge your changes.

Thank you

**Comment by coreywan** [comment]
@lonico ,

I'm okay if you take my commit and merge it into your change.

**Comment by lonico** [comment]
@coreywan https://github.com/ansible/ansible/pull/53993 Let me know if you would like to add a Copyright line, and what you'd like to see in this line.

**Comment by coreywan** [comment]
@lonico, what you put in the merge was more than enough. Thanks! I will close out this PR.

**Pull request closed without merging**